### ISSUE TITLE
Windows 8.1 ndm install node-gyp rebuild failing

### ISSUE THREAD

**Issue by essingh** [agnat/node_mdns#113]
I've installed the Apple Bonjour SDK and VS2013 express.. But when "node-gyp rebuild" executes during the ndm install, there're header files missing (qos.h, ras.h, lmcons.h, mprapi.h and more) causing the build to fail. I was manually putting the headers in c:\Program Files (x86)\Windows Kits\8.1\Include\um and the build would find them, but only to report additional missing headers.. So, question: have you successfully used mdns in Windows 8.1 with VS2013 express? Thanks for you time. 

(nice little module and it works like a charm on the osx version of my app in node webkit)

**Comment by agnat** [comment]
> So, question: have you successfully used mdns in Windows 8.1 with VS2013 express?

I'm afraid I haven't. I currently don't have windows at all. :-/

From what you describe we're probably missing an include directory. Could you undo any manual copying and [gist](http://gist.github.com) the full output of `node-gyp rebuild`? Also, what directory did you take the missing header files from when manually putting them into the `um` directory?

If you don't mind a little debugging ping-pong...

**Comment by y-lohse** [comment]
I have it working on windows 7 with VS2013. This may not be the cause of your problem, but worth a shot: the regular windows terminal is missing the uh.. MSVS context. MSVS2013 ships with it's own command prompt that you should use to compile things.
You also need to `rebuild` using the `--msvs_version=2012` flag. I personnaly run `nw-gyp clean`, `nw-gyp configure --msvs_version=2012` & `nw-gyp build target=0.12.0` or whatever node webkit version you want.

**Comment by essingh** [comment]
y-lohse, I'm actually having another issue with nw-gyp ([gist](https://gist.github.com/essingh/ca337da562b0ac689c60)) that appears to be webkit config specific (can't reach a tar file on amazon aws s3) and something I'll circle-back to and tackle later. Because of that, I decided to just make a new vanilla (not webkit) node project (where mdns was not already installed in my branch) to mess around with mdns itself, which led to where I'm at right now. But thanks for the info because when I was trying to get my branch to work (mdns already installed) I didn't use vs2013 command prompt, but the normal (for lack of a better term) command prompt and I used `--msvs_version=2013` and not 2012. I'll be sure to use your notes when I get that far with webkit; I'm not very familiar with node dev on Windows so there's a learning curve here.

agnat, here's the [gist](https://gist.github.com/essingh/95856edfe24441502788) output from "npm install mdns" in a clean/empty project dir after doing as you requested (undoing the manual copying). As you can see, it's complaining about a missing qos.h header file. And  [here](https://gitorious.org/w64/w64/source/33c44af8cc18cecb8d2feb518b637381fb44e8d9:mingw-w64-headers/include) is where I got that header file and I put it in `C:\Program Files (x86)\Windows Kits\8.1\Include\um`. Once I did that, the build was able to move forward, but than I had the same issue with netioapi.h, iprtrmib.h, mprapi.h, lmcons.h and ras.h. I was able to fine all those headers at the location I mentioned above, however, then it was complaining about not finding _mingw_unicode.h. I googled and found _mingw_unicode.h and added it, but than the build really blew-up with a sea of read, at which point I decided to ask you. Also, I have no idea if those headers are the correct versions etc, I was just crossing my fingers hoping for the best.  I recently installed Apple Bootcamp and Windows on my mac book, so if you don't mind dealing with some of my Windows dev ignorance about all of this, I'd be glad to help with debugging.

**Comment by agnat** [comment]
Hehe, yeah well... no. These are mingw headers and they won't do it.

First, create another gist with the contents of `C:\Program Files (x86)\Windows Kits\8.1\Include\shared`. Look for the missing header files in that directory.

Second, install the [Windows SDK](http://msdn.microsoft.com/en-us/windows/desktop/bg162891.aspx) and re-run `npm install mdns` in a fresh command prompt. [From what I read](http://stackoverflow.com/questions/19736036/visual-studio-2013-fatal-error-c1083-cannot-open-include-file-winsock2-h-n) installing the Windows SDK influences the migration of project files. It might just do the trick.

If not get back me...

**Comment by y-lohse** [comment]
Alright, even though I think I had to use the MSVS prompt to compile mdns with node-gyp too. And I'm not entirely sure of the value to use for msvs_version, I believe there's a shift between the MSVS year and version. Anyway, good luck :)

**Comment by agnat** [comment]
Yeah, well. IIRC gyp calls the `vcvarsall.bat` script, which sets up the VS environment. So, it shouldn't be necessary to use the VS command prompt. However, if it doesn't work for some reason trying the VS command prompt is probably a good idea. Maybe VS2013 does things differently.

**Comment by essingh** [comment]
Ok cool so installing the Windows SDK put the headers in `C:\Program Files (x86)\Windows Kits\8.1\Include\shared` and the build finds them and proceeds. However, then it complains about a missing lib: libcpmt.lib. Here's the [gist](https://gist.github.com/essingh/34bcb783dc874fbd699b) line 143. I did some searching on that and it looks like others had the same issue after installing vs2013 and the Windows SDK. One of the suggestions I found to resolve the issue was to do a repair on vs2013, which I did, but it didn't resolve the missing libcpmt.lib. I'll keep searching. Thanks again for all the help.

**Comment by essingh** [comment]
I found libcpmt.lib (and libcmt.lib which was also missing) in `C:\Program Files (x86)\Microsoft Visual Studio 11.0\VC\lib\amd64` and put it in `C:\Program Files (x86)\Windows Kits\8.1\Lib\winv6.3\um\x64`. This solved the missing lib complaints. However, now it complains the libs are mismatched..here's the [output](https://gist.github.com/essingh/2e73b37f226b10ab0901). What little searching I did on the subject ([here](http://stackoverflow.com/questions/19575747/error-lnk2038-mismatch-detected-for-msc-ver-value-1600-doesnt-match-valu)) it sounds like these libs need to be recreated with vs2013, but I'm not positive. If a recompile in vs2013 is indeed required, is it reasonable to say that mdns isn't quite ready to be used with vs2013?

**Comment by agnat** [comment]
Nah, don't fret. This is all rather typical. Just let me get a branch going real quick. In the meantime: Please undo any manual copying. This is _not_ how we do things.... ;)
